### ISSUE TITLE
YubiKey Plus not recognized

### ISSUE THREAD

**Issue by roycewilliams** [Yubico/yubikey-manager-qt#274]
Even after #247, with the latest release, the "No YubiKey present" error is produced upon insertion of a YubiKey Plus (idVendor=1050, idProduct=0410). Reproducible with three different units.

**Comment by dainnilsson** [comment]
The changes in #247 should be cosmetic only (in the selection of product image). The YubiKey Plus *should* work just fine, both before and after this change. It sounds like there might be a permissions issue instead, preventing YubiKey Manager from accessing the keys. Which OS are you running?

**Comment by roycewilliams** [comment]
Ubuntu 20.04. My original PR related to that ticket was tested and working to add support for an existing key:

https://github.com/Yubico/yubikey-manager-qt/pull/247/commits/8d38d18e4d64a1d01d7a7b215763759e461353d5

... because only the detection function, matching the name as reported by the key, was missing.

Perhaps only the cosmetic pieces were adapted from my PR?

If it was a permissions / udev /etc issue, I would expect all key types to be affected. I have a variety of other YubiKeys and they're detected fine - only the YubiKey Plus is affected.

**Comment by dainnilsson** [comment]
Can you run the command line tool's diagnostics command, `ykman --diagnose` (with the YubiKey Plus attached), and provide the output of that here? You'll need ykman 4.0.0 or later (4.0.1 is the current release) and you should be able to get it from our PPA:

```
sudo add-apt-repository ppa:yubico/stable
sudo apt update
sudo apt install yubikey-manager
```

**Comment by roycewilliams** [comment]
Done - apologies for the delay.

First, the YubiKey Plus:

```
ykman: 4.0.1
Python: 3.8.5 (default, Jan 27 2021, 15:41:15) 
[GCC 9.3.0]
Platform: linux
Running as admin: False

Detected PC/SC readers:

Detected YubiKeys over PC/SC:

Detected YubiKeys over HID OTP:
	OtpYubiKeyDevice(pid=0410, fingerprint='/dev/hidraw5')
	Failed to read device info: No data
	OTP: ConfigState(configured: (True, False), touch_triggered: (False, False), led_inverted: False)


Detected YubiKeys over HID FIDO:
	CtapYubiKeyDevice(pid=0410, fingerprint='/dev/hidraw6')
CTAP device version: 4.0.0
CTAPHID protocol version: 2
Capabilities: 1
	Failed to read device info: CTAP error: 0x01 - INVALID_COMMAND
	Couldn't get info: Device does not support CTAP2.

End of diagnostics
```


Second, a Security Key NFC:

```
ykman: 4.0.1
Python: 3.8.5 (default, Jan 27 2021, 15:41:15) 
[GCC 9.3.0]
Platform: linux
Running as admin: False

Detected PC/SC readers:

Detected YubiKeys over PC/SC:

Detected YubiKeys over HID OTP:

Detected YubiKeys over HID FIDO:
	CtapYubiKeyDevice(pid=0120, fingerprint='/dev/hidraw5')
CTAP device version: 5.1.2
CTAPHID protocol version: 2
Capabilities: 5
	DeviceInfo(config=DeviceConfig(enabled_capabilities={<TRANSPORT.USB: 'usb'>: <CAPABILITY.FIDO2|U2F: 514>, <TRANSPORT.NFC: 'nfc'>: <CAPABILITY.FIDO2|U2F: 514>}, auto_eject_timeout=0, challenge_response_timeout=15, device_flags=<DEVICE_FLAG.0: 0>), serial=None, version=Version(major=5, minor=1, patch=2), form_factor=<FORM_FACTOR.USB_A_KEYCHAIN: 1>, supported_capabilities={<TRANSPORT.USB: 'usb'>: <CAPABILITY.FIDO2|U2F: 514>, <TRANSPORT.NFC: 'nfc'>: <CAPABILITY.FIDO2|U2F: 514>}, is_locked=False, is_fips=False)
	RawInfo: 25010202020302020204010105030501020602000007010f0801000d0202020e0202020a0100
	Ctap2Info: {<VERSIONS: 0x01>: ['U2F_V2', 'FIDO_2_0'], <EXTENSIONS: 0x02>: ['hmac-secret'], <AAGUID: 0x03>: b'mD\xba\x9b\xf6\xec.I\xb90\x0c\x8f\xe9 \xcbs', <OPTIONS: 0x04>: {'rk': True, 'up': True, 'plat': False, 'clientPin': True}, <MAX_MSG_SIZE: 0x05>: 1200, <PIN_UV_PROTOCOLS: 0x06>: [1]}
PIN retries: (8, None)

End of diagnostics
```

**Comment by roycewilliams** [comment]
If some model detection is based on string matching for the USB device type ... would 'plus' being lower-case be a problem?
```
Bus 002 Device 069: ID 1050:0410 Yubico.com Yubikey plus OTP+U2F
Device Descriptor:
  bLength                18
  bcdUSB               2.00
  bMaxPacketSize0        64
  idVendor           0x1050 Yubico.com
  idProduct          0x0410 Yubikey plus OTP+U2F
```
... via:
```
  $ grep -i 'yubikey plus' /usr/share/misc/usb.ids 
	0410  Yubikey plus OTP+U2F
```

**Comment by dainnilsson** [comment]
Thanks!

Based on the output it looks like ykman is able to access the device correctly using both the OTP and FIDO USB endpoints (so no permissions issue). Do other ykman commands work with the Plus? For example, what is the output of `ykman info`? If that looks correct, you can also try running the GUI with debug logging enabled by adding `--log-level debug` when running the GUI.

The USB descriptor "product name" isn't used by the tools for this at all, so that shouldn't have nay impact.

**Comment by roycewilliams** [comment]
Yep, given that I can use all three of my Plus keys for U2F, and that they were detected fine with the patch that I submitted above, I'm pretty sure that the keys themselves are fine.

```
$ ykman info
Device type: YubiKey Plus
Firmware version: 3.0.0
Form factor: Keychain (USB-A)
Enabled USB interfaces: OTP, FIDO

Applications
FIDO2   	Not available	
OTP     	Not available	
FIDO U2F	Enabled      	
OATH    	Not available	
OpenPGP 	Not available	
PIV     	Not available	
```

**Comment by dainnilsson** [comment]
Can you try running the GUI with `--log-level debug` as well? All the output from the CLI commands looks absolutely correct, so it does look very likely that it's something in the GUI (which is strange, because just about all the relevant handling is in the CLI portion).

I've looked at your patch again, and there is nothing in it that looks like it would make any functional difference besides displaying the correct image (VS a more generic YubiKey image), which is exactly what we added in the 1.2.0 release. I can't figure out why applying that patch would make any difference to being able to detect the key in question.

**Comment by roycewilliams** [comment]
Ah, 'DEBUG' works much better than 'debug' - thanks!

Plus:
```
$ ./yubikey-manager-qt-1.2.1-linux.AppImage --log-level DEBUG
2021-04-08T12:22:29-0800 INFO [ykman.logging_setup.setup:74] Initialized logging for level: DEBUG
2021-04-08T12:22:29-0800 INFO [ykman.logging_setup.setup:75] Running ykman version: 4.0.1
2021-04-08T12:22:29-0800 DEBUG [ykman.logging_setup.log_sys_info:47] Python: 3.8.3 (default, Mar 30 2021, 08:11:11) 
[GCC 5.4.0 20160609]
2021-04-08T12:22:29-0800 DEBUG [ykman.logging_setup.log_sys_info:48] Platform: linux
2021-04-08T12:22:29-0800 DEBUG [ykman.logging_setup.log_sys_info:54] Running as admin: False
2021-04-08T12:23:04-0800 DEBUG [yubikit.core.otp.send_and_receive:160] SEND: 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000106b5b000000
2021-04-08T12:23:04-0800 DEBUG [yubikit.core.otp.send_and_receive:164] RECV: 003d0b98667900
2021-04-08T12:23:04-0800 DEBUG [ykman.device._read_info_otp:313] Unable to get info via Management application, use fallback
2021-04-08T12:23:04-0800 DEBUG [yubikey.refresh:138] Sleep...
2021-04-08T12:23:05-0800 DEBUG [yubikit.core.otp.send_and_receive:160] SEND: 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000106b5b000000
2021-04-08T12:23:05-0800 DEBUG [yubikit.core.otp.send_and_receive:164] RECV: 003d0b98667900
2021-04-08T12:23:05-0800 DEBUG [ykman.device._read_info_otp:313] Unable to get info via Management application, use fallback
2021-04-08T12:23:05-0800 DEBUG [yubikey.refresh:138] Sleep...
2021-04-08T12:23:05-0800 DEBUG [yubikit.core.otp.send_and_receive:160] SEND: 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000106b5b000000
2021-04-08T12:23:05-0800 DEBUG [yubikit.core.otp.send_and_receive:164] RECV: 003d0b98667900
2021-04-08T12:23:05-0800 DEBUG [ykman.device._read_info_otp:313] Unable to get info via Management application, use fallback
2021-04-08T12:23:06-0800 DEBUG [yubikit.core.otp.send_and_receive:160] SEND: 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000106b5b000000
2021-04-08T12:23:06-0800 DEBUG [yubikit.core.otp.send_and_receive:164] RECV: 003d0b98667900
```

Security Key NFC:
```
2021-04-08T12:23:51-0800 DEBUG [ykman.device.read_info:393] Read info: DeviceInfo(config=DeviceConfig(enabled_capabilities={<TRANSPORT.USB: 'usb'>: <CAPABILITY.FIDO2|U2F: 514>, <TRANSPORT.NFC: 'nfc'>: <CAPABILITY.FIDO2|U2F: 514>}, auto_eject_timeout=0, challenge_response_timeout=15, device_flags=<DEVICE_FLAG.0: 0>), serial=None, version=Version(major=5, minor=1, patch=2), form_factor=<FORM_FACTOR.USB_A_KEYCHAIN: 1>, supported_capabilities={<TRANSPORT.USB: 'usb'>: <CAPABILITY.FIDO2|U2F: 514>, <TRANSPORT.NFC: 'nfc'>: <CAPABILITY.FIDO2|U2F: 514>}, is_locked=False, is_fips=False)
```

**Comment by dainnilsson** [comment]
Thanks! I should have spotted it sooner, but your `ykman info` output shows `OTP     	Not available`, which is clearly wrong. This probably means that the issue that is causing this is in ykman CLI. I'm going to do some more digging...

**Comment by dainnilsson** [comment]
Aaaaaaand bingo! Here's the facepalming bug that's causing this: https://github.com/Yubico/yubikey-manager/commit/614688dff254f7577a0cd263cc73df91bfa362c5

Again, thanks for the help in debugging! We expect to have an actual YubiKey Plus device in our hands later today so that we can verify for sure. Assuming all goes well we'll release new patch versions of the tools next week.

**Comment by fdennis** [comment]
@roycewilliams We have created a build that contains a fix for the bug mentioned above. We would appreciate it if you could help us by testing if it now properly recognizes your key. You can find the build here https://github.com/Yubico/yubikey-manager-qt/actions/runs/733698345 that contains the latest AppImage (not yet released) at the bottom of the page, under "Artifacts". It should download a .zip file that you just unpack.

Let us know if you run into any issues.

**Comment by roycewilliams** [comment]
@fdennis Thanks! Zip unpacked cleanly.

Same results so far. Debug:

```
$ sha256sum yubikey-manager-qt-actions-gpg-mac.AppImage
e757eb5e5702f8c7a3202cdac944eb32aa896a9c3ad59e2c745411c81506c0a4  yubikey-manager-qt-actions-gpg-mac.AppImage
$ ./yubikey-manager-qt-actions-gpg-mac.AppImage --log-level DEBUG
2021-04-09T12:48:05-0800 INFO [ykman.logging_setup.setup:74] Initialized logging for level: DEBUG
2021-04-09T12:48:05-0800 INFO [ykman.logging_setup.setup:75] Running ykman version: 4.0.1
2021-04-09T12:48:05-0800 DEBUG [ykman.logging_setup.log_sys_info:47] Python: 3.8.3 (default, Apr  9 2021, 16:37:17) 
[GCC 5.4.0 20160609]
2021-04-09T12:48:05-0800 DEBUG [ykman.logging_setup.log_sys_info:48] Platform: linux
2021-04-09T12:48:05-0800 DEBUG [ykman.logging_setup.log_sys_info:54] Running as admin: False
2021-04-09T12:48:14-0800 DEBUG [yubikit.core.otp.send_and_receive:160] SEND: 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000106b5b000000
2021-04-09T12:48:14-0800 DEBUG [yubikit.core.otp.send_and_receive:164] RECV: 003d0b98667900
2021-04-09T12:48:14-0800 DEBUG [ykman.device._read_info_otp:313] Unable to get info via Management application, use fallback
2021-04-09T12:48:14-0800 DEBUG [yubikey.refresh:138] Sleep...
[repeats]
```

**Comment by dainnilsson** [comment]
I've checked the sha256 sum and it looks correct. That AppImage contains the newer ykman version which should have the fix. However, that should show `Running ykman version: 4.0.2-dev0` while yours is still showing `4.0.1`. I just tested now and confirmed that the AppImage picks up ykman from the system rather than using the embedded version (I'll open a separate issue to deal with that).

Unfortunately we were unable to fins a YubiKey Plus device ourselves (we thought we had one at the office, but it turned out we didn't). If you'll bear with us for just a bit longer, there is a way to test the AppImage using the correct ykman version:

Open a terminal and cd into the directory where your AppImage file is, then:
```
$ virtualenv ykman-venv
$ source ykman-venv/bin/activate
(ykman-venv) $ pip install "git+https://github.com/Yubico/yubikey-manager.git@main"
(ykman-venv) $ ./yubikey-manager-qt-actions-gpg-mac.AppImage --log-level DEBUG
```

This will create a new virtualenv, install ykman from the latest git commit, and run the AppImage in that environment. You should be able to verify that the correct version of ykman is being used by the log output of the AppImage.

**Comment by roycewilliams** [comment]
Thanks - this worked!
```
2021-04-10T07:52:10-0800 INFO [ykman.logging_setup.setup:75] Running ykman version: 4.0.2-dev0
```
(In the meantime, I'd also tried to build the whole stack from source (on Xubuntu), but even after installing the packages in the build doc, the 'poetry' command isn't available. I'm sure it's a noob issue and there's a Google-able way to address that - whatever it is, it should probably be mentioned in the build doc?)

Confirmed that my three Plus keys are now properly detected - with image, firmware version, and serial!

Thank _you_ for taking the time to pursue this case, even though most folks are probably not still rockin' many YubiKey Plus keys in the field. It speaks to Yubico's commitment to long-term support.

**Comment by dainnilsson** [comment]
Great, I'm relieved to hear it! A new version of ykman (CLI) was just released today, and the GUI will be out in another day or so.

Is this the document you're referring to? https://github.com/Yubico/yubikey-manager/blob/main/doc/Development.adoc
We mention requiring poetry to be installed pretty early on, and link to the project page for that. Maybe we can make it more clear that you should refer to the instructions on that site for installing poetry itself!

**Comment by roycewilliams** [comment]
Yep, reading too fast - didn't notice the poetry ref and went straight to the packages list. :) A little more callout for the hurried might be in order. Even making it its own line and converting the contents of the first section to be a bulleted list would probably be enough.

Thanks for the effort for the Plus keys - closing!
